### PR TITLE
Things were badly broken on skybridge.

### PR DESCRIPTION
**Pull request by jgfouca**
Fix list:
1) {{ queue }} and {{ wall_time }} are apparently no longer supported, use job_queue and job_wallclock_time instead.
2) get_project needs to optionally look at config_machines.xml as its documentation advertises
3) Remove methods from batch.py that have been replaced by methods in env_batch.py (cleanup, not fix)
4) env_batch::get_value needs to use attributes when looking for optional node
5) Need to take job and case into account when getting batch directives
6) Fix env_batch::get_default_walltime
7) Fallback to default walltime if queue does not provide max walltime
8) Namespace errors in buildnml and buildlib
9) clone_case should use the PROJECT from the previous case
10) Make names line up better between env_batch and task_maker. This will allow
    easier substitution between the two when resolving {{ tokens }}.
11) Mpi args were not being handled correctly. Do not determine args until an
    mpirun has been selected.
12) empty-dict attributes was breaking get_nodes

Test suite: cime_developer
Test baseline: N/A
Test namelist changes: N/A
Test status: bit for bit

Fixes: [CIME Github issue #]

User interface changes?: no

Code review: @mvertens

### CODE REVIEW

**Comment by rljacob** [comment]
Are these fixes only to help things on skybridge?   I see lots of general purpose infrastructure files being modified.

**Comment by jgfouca** [comment]
Things were very broken, not just on skybridge.

**Comment by jgfouca** [comment]
@mvertens hold off on this for a bit, seeing even more problems in broader testing

**Comment by mvertens** [comment]
@foucar I had just started looking at it. I don't know a lot about the
batch implementation. I can review the rest fairly comfortably.

File: utils/python/CIME/case.py [hunk 1]
```diff
@@ -758,7 +753,7 @@ def create_clone(self, newcase, keepexe=False, mach_dir=None, project=None):
         # user's case. However, note that, if a project is not given, the fallback will
         # be to copy it from the clone, just like other xml variables are copied.
         if project is None:
-            project = get_project()
+            project = self.get_value("PROJECT")
```
Comment (quantheory) [inline utils/python/CIME/case.py]: Is this change just to make sure that the code is doing what the comment actually says it's doing?

Comment (jgfouca) [inline utils/python/CIME/case.py]: Yes.

**Comment by quantheory** [comment]
This looks good to me for the most part. Regarding `ccsm_estcost`, the answer is yes, it should be taken into account (especially for regression tests where you don't want to have to monkey with each case by hand).

I think what should be happening is that, instead of using the max wall clock time, there should be a function that searches the list of walltimes and selects the one that has `ccsm_estcost` as small as possible while still at least as big as the current value of `CCSM_ESTCOST` in the case. Only if there's no such walltime listed should it fall back on the `walltimemax` setting. And if that's missing should we use the default, as you implemented.

**Comment by rljacob** [comment]
I don't understand your second paragraph about ccsm_estcost.  Is it different from CCSM_ESTCOST?

We would like to set an individual walltime for each test because they could get through the queue faster if the walltime requested was what was needed and smaller then the max wall clock time.

**Comment by quantheory** [comment]
Well, `CCSM_ESTCOST` is in the case (`env_mach_pes`), and specifies the estimated cost of that case, but there's also a `ccsm_estcost` attribute on the `walltime` elements in `config_machines`, and my understanding is that that's supposed to represent approximately how long a run with that estimated cost will take. It's a pretty crude way of matching things up, since AFAIK nothing in the cost calculations accounts for how many days you are running? Or the number of cores used? But it should at least be better than just using the max wall time every single time.

**Comment by mvertens** [comment]
I can see still needed ccsm_estcost - but I think it should be renamed estcost - and get rid of the ccsm_ prefix. But this has nothing to do with this particular PR. However, do we still need CCSM_ESTCOST?

**Comment by mvertens** [comment]
This looks fine to me. Could you please first fill in the commit template - and I'll then go ahead and merge the PR.

**Comment by quantheory** [comment]
LGTM

@mvertens `CCSM_ESTCOST` is the actual estimated cost, isn't it? `ccsm_estcost` is just a way of translating it that to a walltime?

**Comment by jgfouca** [comment]
OK, found several more problems and fixed them. Skybridge appears to be functional again.

**Comment by jgfouca** [comment]
I'm going to go ahead and merge. I don't think any of my recent fixes are controversial.

**Comment by jedwards4b** [comment]
I have been trying to phase out these estimated cost variables in favor of specifying the wallclock time for tests in testlist.xml.   Not sure where things stand at this point.

**Pull request merged**